A collection of Matt and David's commits from #2316

rwest opened the pull request:
<!--
Thanks for contributing a pull request! Please try to provide as much detail as possible to help the reviewer understand your work.
You can also add the appropriate labels to describe the topic of the pull request and the type of changes you're making.
-->

### Motivation or Problem
Pull request #2316 which adds electrochemistry, adds lithium, adds solvation to surfaces, adds charge transfer reactions, is rather large and hard to review, and also has some necessarry but unrelated things, that can be helpful elsewhere.

### Description of Changes
I've pulled out a bunch (but not all) of those things into this PR so they can be reviewed and merged, then the electrochem PR can be rebased onto it.   I also reordered things so the fixups commits are mostly right after the things they're fixing.

One feature that was added was the ability to forbid species and functional groups in input files. As I was trying to add examples to the input files and documentation, I found it was already there, since this feature was added in https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2185 - authored around the same time in October 2021. Rather than add a second syntax for the same thing, I reverted the commits on this branch.

### Testing
I read the commits as I was cherry-picking, got the test suite to pass, and also added some more unit tests.

<!--
### Reviewer Tips
Suggestions for verifying that this PR works or other notes for the reviewer.
Checklist before submission:
 - Have you added appropriate unit tests?
 - Have you checked that all unit tests pass?
 - Is your code commented and understandable?
 - Have you updated related documentation?
 - Are the commits logically organized and informative?
 - Is your branch up to date with main?
-->

rwest commented:
The only test failures are expected failures of the regression tests because of the random run-to-run variations.

rwest commented on rmgpy/data/kinetics/family.py:
```diff
@@ -4670,3 +4673,80 @@ def _child_make_tree_nodes(family, child_conn, template_rxn_map, obj, T, nprocs,
+    else:
+        raise Exception('Invalid units {0} for averaging kinetics.'.format(Aunits))
+
+    if type(kinetics) not in [Arrhenius,SurfaceChargeTransfer,ArrheniusChargeTransfer]:
```
Oops. These `ChargeTransfer` reaction classes not yet defined on this branch (a result of cherry-picking out of order)

rwest replied:
I removed those (in a commit that can be reverted when needed)

rwest commented on rmgpy/rmg/input.py:
```diff
@@ -1526,7 +1528,8 @@ def read_input_file(path, rmg0):
         'InChI': inchi,
         'adjacencyList': adjacency_list,
-        'adjacencyListGroup': adjacency_list_group,
+        'adjacencyListGroup': group_adjacency_list,
```
We should add to one of the example input files, and documentation.

rwest replied:
Instead I removed this feature, because it already exists in https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2185

JacksonBurns commented on rmgpy/data/kinetics/family.py:
```diff
@@ -4670,3 +4673,80 @@ def _child_make_tree_nodes(family, child_conn, template_rxn_map, obj, T, nprocs,
+        if electrons is None:
+            electrons = kinetics_list[0].electrons.value_si
+        assert all(np.abs(k.V0.value_si) < 0.0001 for k in kinetics_list), [k.V0.value_si for k in kinetics_list]
+        assert all(np.abs(k.alpha.value_si - 0.5) < 0.001 for k in kinetics_list), [k.alpha for k in kinetics_list]
```
Could these `assert` statements include a message, i.e. `assert all(...), "Expected all V0 to be..." - right now this will raise odd errors to the user, should these asserts not pass.

rwest replied:
I replaced the Asserts with ValueErrors

rwest reviewed (commented):
@mjohnson541 isn't this "make it possible to explicitly forbid molecules and groups in the input file"  feature already here thanks to https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2185 ? 

Besides adding a new, undocumented, name for it, not sure what this commit adds? Can you use the existing mechanism in your input files and we drop this commit?

I see you authored this commit a couple of weeks before PR #2185 was merged, so you're forgiven for not using it back then :)

JacksonBurns commented on rmgpy/data/kinetics/family.py:
```diff
@@ -4670,3 +4673,55 @@ def _child_make_tree_nodes(family, child_conn, template_rxn_map, obj, T, nprocs,
+    n /= count
+    Ea /= count
+    Aunits = kinetics_list[0].A.units
+    if Aunits == 'cm^3/(mol*s)' or Aunits == 'cm^3/(molecule*s)' or Aunits == 'm^3/(molecule*s)':
```
All of these statements should look like `if Aunits in ("cm^3/(mol*s), (cm^3/(molecule*s)), ...)` rather than the big or piles.

JacksonBurns replied:
Note use of tuple, not list.

rwest replied:
I was curious...
<img width="869" alt="Screenshot 2023-07-14 at 2 09 45 PM" src="https://github.com/ReactionMechanismGenerator/RMG-Py/assets/93807/e9f70767-ead7-4ef9-b102-f1754e27357b">

JacksonBurns replied:
This isn't a speed thing as much as an overly strict linting rule - see [pylint's R1714](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/consider-using-in.html) - that I just happen to really like. I was actually wrong though, we should be using a set which should run even faster.

rwest replied:
You're right that `in set` is faster still.

JacksonBurns reviewed (changes requested):
Looks good overall, have requested some small changes for good code style and clarity.

rwest commented on rmgpy/data/kinetics/family.py:
```diff
@@ -4670,3 +4673,55 @@ def _child_make_tree_nodes(family, child_conn, template_rxn_map, obj, T, nprocs,
+    logA = 0.0
+    n = 0.0
+    Ea = 0.0
+    count = 0
+    for kinetics in kinetics_list:
+        count += 1
+        logA += np.log10(kinetics.A.value_si)
+        n += kinetics.n.value_si
+        Ea += kinetics.Ea.value_si
+
+    logA /= count
+    n /= count
+    Ea /= count
```
But we'd have to iterate through the list multiple times (once for each attribute) and create a bunch of lists. Is this a verbosity thing? an efficiency thing? a less-likely-to-make-mistakes thing?
I don't like playing code golf for the sake of it.

JacksonBurns replied:
> But we'd have to iterate through the list multiple times (once for each attribute) and create a bunch of lists. 

You can do it very efficiently in one line like this:
```python
from types import SimpleNamespace
import numpy as np

# mimic the kinetics objects
kinetics_list = [SimpleNamespace(a=1, b=2, c=3) for _ in range(3)]

a, b, c = np.mean([[np.log10(i.a), i.b, i.c] for i in kinetics_list], axis=1)
```

> Is this a verbosity thing? an efficiency thing? a less-likely-to-make-mistakes thing? I don't like playing code golf for the sake of it.

This is an antipattern, which we benefit from removing for all of these reasons. Numpy will be faster, less code is easier to maintain and read, etc.

rwest replied:
A hole in one!

I quite like it. Although, having now _taught_ programming a few times, I am learning to appreciate the benefits of code that a novice can understand at a glance. More participation, fewer mistakes, less time spent explaining or parsing, etc. etc.

Anyway, the main reason to not do it your way here and now is that we soon need to revert this commit 
https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2491/commits/21801750f0396821f5700f0a6a344adecf3c927f
and the merge conflicts would be horrible to resolve.

We can revisit optimizing this code after that happens. 
(Though I think this code is seldom called)

rwest commented on arkane/common.py:
```diff
@@ -390,7 +390,11 @@ def get_element_mass(input_element, isotope=None):
     elif isinstance(input_element, str):
         symbol = input_element
-        number = next(key for key, value in symbol_by_number.items() if value == input_element)
+        try:
```
Done in cfe970f4bcaf4037696c7625f1439811c3db678f

rwest commented:
Think I've addressed all your comments. I pushed back on one suggestion - though I like it - because it'll cause horrible merge conflicts soon because that method needs to be more complicated than it currently looks. But the rest I adopted.

JacksonBurns reviewed (approved):
LGTM!
